source/s3_request.c

```diff
@@ -59,6 +75,19 @@ void aws_s3_request_clean_up_send_data(struct aws_s3_request *request) {
 
     aws_signable_destroy(request->send_data.signable);
     request->send_data.signable = NULL;
+    if (request->send_data.metrics) {
+        /* invoke callback */
+        struct aws_s3_meta_request *meta_request = request->meta_request;
+        struct aws_s3_request_metrics *metric = request->send_data.metrics;
+        aws_high_res_clock_get_ticks(&metric->time_metrics.end_timestamp_ns);
+        metric->time_metrics.total_duration_ns =
+            metric->time_metrics.end_timestamp_ns - metric->time_metrics.start_timestamp_ns;
+        /* End the timestamp */
+        if (meta_request->telemetry_callback) {
+            meta_request->telemetry_callback(meta_request, metric, meta_request->user_data);
```

Comment:
Is this the right place to trigger this callback? We also call this function in `aws_s3_request_setup_send_data`

Reply:
I think yes.

https://github.com/awslabs/aws-c-s3/blob/main/include/aws/s3/private/s3_request.h#L84

The send_data should be around for each time the request is sent, so, if any error or retries happens for the aws_s3_request, we will still get the metrics for that time the real http request sent/to be sent.